### DISCUSSION TITLE
pytest --dashboard does not generate dashboard.html in the directory

### DISCUSSION THREAD

**Discussion by LiPingYen** [seleniumbase/SeleniumBase#2296] [Q&A] [answered]
Hi everyone,
I ran `pytest test.py -n4 --dashboard --html=report.html` in pipenv, but during the execution, I couldn't find dashboard.html in the target directory. Has anyone encountered the same situation, and if so, is there a corresponding solution? Thanks.

**Comment by mdmintz** [comment, marked as answer]
This is what I saw when running it:

```
SeleniumBase/examples michael$ pytest test_repeat_tests.py -n4 --dashboard --html=report.html
============================== test session starts ===============================
platform darwin -- Python 3.12.0, pytest-7.4.3, pluggy-1.3.0
rootdir: /Users/michael/github/SeleniumBase/examples
configfile: pytest.ini
plugins: html-2.0.1, rerunfailures-12.0, xdist-3.4.0, cov-4.1.0, metadata-3.0.0, ordering-0.6, seleniumbase-4.21.4
collecting: 1/4 workers 
Dashboard: file:///Users/michael/github/SeleniumBase/examples/dashboard.html
****************************************************************************
4 workers [6 items]    
......
- generated html file: file:///Users/michael/github/SeleniumBase/examples/report.html -
```

The dashboard link should be displayed in the console. If not, add `-s` as a `pytest` command-line option if you weren't using the `pytest.ini` file provided: [SeleniumBase/examples/pytest.ini](https://github.com/seleniumbase/SeleniumBase/blob/master/examples/pytest.ini)

**Reply by LiPingYen** [reply]
Thanks Michael,

I ran the code again as below
```
pytest test_get_one_tour_photo_only_SB_final_version.py -n4 --dashboard --html=report.html --rs --crumbs -s
Test session starts (platform: darwin, Python 3.11.5, pytest 7.4.3, pytest-sugar 0.9.7)
rootdir: /Users/taylor.yen/Desktop/Klook/a_and_t_project/data_science_implement/aid_mapping_scraping
plugins: html-2.0.1, anyio-4.0.0, rerunfailures-12.0, xdist-3.4.0, seleniumbase-4.21.4, metadata-3.0.0, ordering-0.6, sugar-0.9.7
ready: 4/4 workers      collecting ... 
Dashboard: file:///Users/taylor.yen/Desktop/dashboard.html
****************************************************************************************************************************
4 workers [32677 items]

 test_get_one_tour_photo_only_SB_final_version.py ✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓
```

The link showed in the terminal, but the dashboard.html file still did 
not show in my Desktop.
Do you have any idea how to fix this?

**Reply by mdmintz** [reply]
That's odd that your `rootdir` is `/Users/taylor.yen/Desktop/Klook/a_and_t_project/data_science_implement/aid_mapping_scraping` and yet your `dashboard` is at `file:///Users/taylor.yen/Desktop/`. If anything, your `rootdir` should be higher, not lower, than where you initiate tests from.

You should probably create a `tests/` folder in `/Users/taylor.yen/Desktop/Klook/a_and_t_project/data_science_implement/aid_mapping_scraping`, and then call `pytest` from there so that your `dashboard` gets generated in that `tests/` folder.

**Reply by LiPingYen** [reply]
I created a `test/` folder and `pytest` the script once again, and I got the same outcome that the dashboard.html is still not in the `test/` folder.

The following image showed what in the `test/`.

```
pytest test_get_one_tour_photo_only_SB_final_version.py -n4 --dashboard --html=report.html --rs --crumbs -s
Test session starts (platform: darwin, Python 3.11.5, pytest 7.4.3, pytest-sugar 0.9.7)
rootdir: /Users/taylor.yen/Desktop/Klook/a_and_t_project/data_science_implement/aid_mapping_scraping/test
plugins: html-2.0.1, anyio-4.0.0, rerunfailures-12.0, xdist-3.4.0, seleniumbase-4.21.4, metadata-3.0.0, ordering-0.6, sugar-0.9.7
ready: 4/4 workers      collecting ... 
Dashboard: file:///Users/taylor.yen/Desktop/Klook/a_and_t_project/data_science_implement/aid_mapping_scraping/test/dashboard.html
*********************************************************************************************************************************
4 workers [32677 items]

 test_get_one_tour_photo_only_SB_final_version.py ✓✓✓✓✓✓✓✓   
```

<img width="974" alt="Unknown" src="https://github.com/seleniumbase/SeleniumBase/assets/47491041/f5933eaf-9ba6-47ce-a7e6-8ce4b450eb13">

**Reply by mdmintz** [reply]
Inside [SeleniumBase/seleniumbase/fixtures/base_case.py](https://github.com/seleniumbase/SeleniumBase/blob/master/seleniumbase/fixtures/base_case.py) there's a section of code that creates the dashboard:

```python
        abs_path = os.path.abspath(".")
        file_path = os.path.join(abs_path, "dashboard.html")
        out_file = codecs.open(file_path, "w+", encoding="utf-8")
        out_file.writelines(the_html)
        out_file.close()
```

I'm not sure what's happening in your scenario, but try some Python debugging with `breakpoint()` to see why that's not creating the dashboard for you. Also see if the same issue happens when you're not using multithreading.

I'll probably be away until Monday night (going on a trip starting this weekend). See if you can figure out the issue before I'm back.

**Reply by LiPingYen** [reply]
Thanks Michael,
I will test and see if the dashboard show up or not.
Thanks again for answering the issue, have a nice trip.

**Comment by LiPingYen** [comment]
After some test, I found out that the dashboard.html only shows when the function is warpped within class.
 
```python
class Test_get_one_tour_photo_only_SB:

    @pytest.mark.parametrize("one_tour_info_list", tour_url_dta_list)

    def test_get_one_tour_photo_only_SB(self ,sb, one_tour_info_list, headless = True):
        
        with SB(uc = True, headless = headless) as sb:
    
            sb.open(tour_main_page_url)
```

If the function aren't wrapped within the class, the dashboard.html won't show.

When testing, some issue related to dashboard showed.
The dashboard only update when the first task completed in each thread.
The dashboard act weirdly both in single thread and multithread task.

**Reply by mdmintz** [reply]
It looks like you're mixing [Syntax Formats](https://github.com/seleniumbase/SeleniumBase/blob/master/help_docs/syntax_formats.md).
The Dashboard only supports `pytest` formats.
The `SB()` format is a pure `python` format.

You can use `--uc` to activate UC Mode with the `pytest` formats.
See [SeleniumBase/examples/verify_undetected.py](https://github.com/seleniumbase/SeleniumBase/blob/master/examples/verify_undetected.py) and [SeleniumBase/examples/uc_cdp_events.py](https://github.com/seleniumbase/SeleniumBase/blob/master/examples/uc_cdp_events.py) for examples.

**Comment by LiPingYen** [comment]
After using pure `pytest` syntax, everything ran smoothly.
I truly appreciate Michael's timely assistance and the fantastic repository he created. 
Wishing you a wonderful holiday in Boston and New York!
